File: ts/src/cmd/clone.ts
import { RemoteClient } from '../remote';
import { RadStore } from '../store';
import { keypairFromSecret, formatPublicKey } from '../crypto';
import * as fs from 'fs';
import * as path from 'path';

export async function runClone(url: string, participant: string, secretKey: string): Promise<string> {
  const client = new RemoteClient(url);

  // Generate public key from secret key
  const kp = keypairFromSecret(secretKey);
  const publicKey = formatPublicKey(kp);

  // 1. Join the relay
  await client.join(participant, publicKey, false);

  // 2. Get all participants
  const participants = await client.getParticipants();

  // 3. Get all operations
  const operations = await client.getLog();

  // 4. Create .rad directory
  const radDir = path.join('.', '.rad');
  fs.mkdirSync(radDir, { recursive: true });

  // 5. Determine founder
  const founder = participants.find(p => p.id === 'alice')?.id
    || participants[0]?.id
    || participant;

  // Create config.json
  const config = { founder };
  fs.writeFileSync(
    path.join(radDir, 'config.json'),
    JSON.stringify(config)
  );

  // Open store and save state
  const store = RadStore.open('.');

  // Save participants
  store.saveParticipants(participants);

  // Save operations
  store.saveOplog(operations);

  // Save remote.json
  const lastTimestamp = operations.length > 0
    ? Math.max(...operations.map(o => o.timestamp))
    : 0;

  const remoteConfig = {
    url,
    lastPullTimestamp: lastTimestamp,
  };

  fs.writeFileSync(
    path.join(radDir, 'remote.json'),
    JSON.stringify(remoteConfig)
  );

  // Build response
  let output = '';
  output += `cloned: rad project from ${url}\n`;
  output += `participants: ${participants.length}\n`;
  output += `operations: ${operations.length}\n`;

  // Count unique files
  const files = new Set(
    operations
      .map(o => {
        const colonPos = o.regionId.indexOf(':');
        return colonPos !== -1 ? o.regionId.substring(0, colonPos) : null;
      })
      .filter(f => f !== null)
  );
  output += `files: ${files.size}\n`;

  return output;
}

File: rust/cli/src/cmd/clone.rs
use crate::remote::RemoteClient;
use crate::store::RadStore;
use crate::crypto;
use std::fs;
use std::path::Path;

pub fn run_clone(url: &str, participant: &str, secret_key: &str) -> Result<String, String> {
    let client = RemoteClient::new(url);

    // Generate public key from secret key
    let kp = crypto::keypair_from_secret(secret_key);
    let public_key = crypto::format_public_key(&kp);

    // 1. Join the relay
    client.join(participant, &public_key, false)?;

    // 2. Get all participants
    let participants = client.get_participants()?;

    // 3. Get all operations
    let operations = client.get_log(None)?;

    // 4. Create .rad directory
    let rad_dir = Path::new(".rad");
    fs::create_dir_all(rad_dir)
        .map_err(|e| format!("Failed to create .rad directory: {}", e))?;

    // 5. Initialize RadStore and save state
    let cwd = std::env::current_dir()
        .map_err(|e| format!("Failed to get current directory: {}", e))?;

    // Create config.json with founder (if we can determine it)
    let founder = if let Some(p) = participants.iter().find(|p| p.id == "alice") {
        p.id.clone()
    } else if let Some(p) = participants.first() {
        p.id.clone()
    } else {
        participant.to_string()
    };

    let config = serde_json::json!({
        "founder": founder,
    });
    fs::write(
        rad_dir.join("config.json"),
        serde_json::to_string(&config).unwrap()
    ).map_err(|e| format!("Failed to write config.json: {}", e))?;

    // Open store
    let store = RadStore::open(&cwd)?;

    // Save participants
    store.save_participants(&participants)?;

    // Save operations
    let mut oplog = crate::oplog::OpLog::new();
    for op in &operations {
        oplog.add_operation(op.clone());
    }
    store.save_oplog(&oplog)?;

    // Save remote.json
    let remote_config = serde_json::json!({
        "url": url,
        "lastPullTimestamp": if operations.is_empty() {
            0
        } else {
            operations.iter().map(|o| o.timestamp).max().unwrap_or(0)
        },
    });
    fs::write(
        rad_dir.join("remote.json"),
        serde_json::to_string(&remote_config).unwrap()
    ).map_err(|e| format!("Failed to write remote.json: {}", e))?;

    // Build response
    let mut output = String::new();
    output.push_str(&format!("cloned: rad project from {}\n", url));
    output.push_str(&format!("participants: {}\n", participants.len()));
    output.push_str(&format!("operations: {}\n", operations.len()));

    // Count unique files from operations
    let files: std::collections::HashSet<String> = operations.iter()
        .filter_map(|o| {
            if let Some(colon_pos) = o.region_id.find(':') {
                Some(o.region_id[..colon_pos].to_string())
            } else {
                None
            }
        })
        .collect();
    output.push_str(&format!("files: {}\n", files.len()));

    Ok(output)
}
